lib/engine/game/g_1841/game.rb

```diff
@@ -363,6 +413,88 @@ def place_home_token(corporation)
           @graph.clear
         end
 
+        # from https://www.redblobgames.com/grids/hexagons
+        def doubleheight_coordinates(hex)
+          [hex.id[1..-1].to_i, hex.id[0].ord - 'A'.ord] # works because AZ isn't close to a home city
+        end
+
+        def hex_distance(hex_a, hex_b)
+          x_a, y_a = doubleheight_coordinates(hex_a)
+          x_b, y_b = doubleheight_coordinates(hex_b)
+
+          # from https://www.redblobgames.com/grids/hexagons#distances
+          # this game essentially uses double-height coordinates
+          dx = (x_a - x_b).abs
+          dy = (y_a - y_b).abs
+          distance = hex_a == hex_b ? -1 : [0, dx + [0, (dy - dx) / 2].max - 1].max
+          distance + 1
+        end
+
+        def tile_laid_at_exit?(hex)
+          hex.tile.exits.any? do |e|
+            next if hex.tile.borders.any? { |b| b.edge == e && ((b.type == :impassable) || (b.type == :province)) }
+
+            neighbor = hex_neighbor(hex, e)
+            neighbor && !neighbor.tile.preprinted
+          end
+        end
+
+        def hex_connected?(hex)
+          !hex.tile.preprinted ||                                  # any tile has been laid here
+            (!hex.tile.paths.empty? && !hex.tile.cities.empty?) || # pre-printed with track and a city
+            (!hex.tile.paths.empty? && tile_laid_at_exit?(hex)) # pre-printed with track and tile connecting to it
+        end
+
+        def search_hexes(current_hex, hex_list, start_hex)
```

Comment:
If I'm understanding correctly, this `search_hexes` will need to be called for token prices to be computed whenever a home token is placed. Is there anything that can be done to mitigate this computation when loading the game, e.g., auto actions for 1822 destination tokens?

Reply:
I thought about optimizing it, but:
1. it's really not very computationally expensive. It's only traversing a maximum of 19 hexes.
2. It will only happen at most once or twice per corporation the entire game.

If you think it will slow down the game, I should be able to skip it when loading.

Reply:
Actually, only doing when not loading is problematic, since the second home token placement may be skipped depending on where the first was placed and that requires doing the search.

I still stand by my statement that this isn't computationally expensive - it's _much_ faster than building a graph.

Reply:
Ignore the last comment, I forgot this wasn't token placement, but token purchase. I could easily change the code such that the token cost is only calculated when not loading. It means that the log won't show the price - only the UI.

Reply:
And...once again, ignore the last comment. I forgot that the choose action doesn't supply enough information (unless I do something ugly like encode both the number of tokens and the price in a single integer).

So I'm back to my assertion that this isn't a performance hit. Also, the token prices only need to be calculated for non-historical companies (about half of them).

Reply:
Sounds good! Seeing something called `search_hexes` just sent up a red flag in my mind, but since it's a much less big deal than building the whole graph I don't think we need to worry about it. 🙂